[doc] add a minimal documentation for OPA RLS capabilities (#22424)
trinodb/trino: docs/src/main/sphinx/security/opa-access-control.md

* - `opa.policy.uri`
  - The **required** URI for the OPA endpoint, for example,
    `https://opa.example.com/v1/data/allow`.
* - `opa.policy.row-filters-uri`
  - The **optional** URI for fetching row filters - if not set no row filtering

>     will be applied, example, `https://opa.example.com/v1/data/rowFilters`.

Reviewer: ```suggestion
    is applied. For example, `https://opa.example.com/v1/data/rowFilters`.
```

* - `opa.policy.column-masking-uri`

>   - The **optional** URI for fetching column masks - if not set no masking will

Reviewer: ```suggestion
  - The **optional** URI for fetching column masks - if not set no masking
```

>     be applied, for example, `https://opa.example.com/v1/data/columnMask`.

Reviewer: ```suggestion
    is applied. For example, `https://opa.example.com/v1/data/columnMask`.
```

>     `https://opa.example.com/v1/data/columnMask`.

Reviewer 2: CC @mosiac1 - once #21997 is merged this should be updated

  }
}
```
## RowFilters 

> Raw filters allow you to denied a group of users to see specifics rows in a table, based on an OPA policy.
> 
> for these kind of policies to work, make sure to set `opa.policy.row-filters-uri` in the opa-plugin configuration
> 
> here is an example bellow

Reviewer 2: ```suggestion
Row filtering allows Trino to filter out some rows from the result before returning it to the caller, controlling
what data different users can see. This plugin supports retrieving filters from OPA.

For these policies to be enabled, make sure to set `opa.policy.row-filters-uri` in the OPA plugin configuration.

For instance, an OPA policy for row filtering may be implemented as below:
```

Author: sorry pushed a update before seing your proposal :/ 
changed have been re-introduced with commit `a714606`

```
            package trino
            import future.keywords.in
            import future.keywords.if
            import future.keywords.contains

            default allow := true

            table_resource := input.action.resource.table
            is_admin {
              input.context.identity.user == "admin"
            }

            rowFilters contains {"expression": "user_type <> 'customer'"} if {
                not is_admin
                table_resource.catalogName == "sample_catalog"
                table_resource.schemaName == "sample_schema"
                table_resource.tableName == "restricted_table"
            }
```

> the format expected by the plugin is a result (here "rowFilters") that contains an array of "{"expressions":"clause"} where each expression is basicaly a where clause.

Reviewer 2: ```suggestion
The response expected by the plugin is an array of objects, each of them in the format `{"expression":"clause"}`.

Each expression essentially behaves like an additional `WHERE` clause.
```

I would mention two more things:

- It is possible for _multiple_ row filters to be returned for a single OPA request, they will all be applied
- Additionally, each object may contain an `identity` field

The `identity` field allows Trino to evaluate these row filters under a different identity - such that a filter can target a column the requesting user cannot see

Author: realy interesting these identity field :) 
thanks !

Author: proposal applied

 

## Column masking

> Column masking allow you to denied access to a particular column instead of the entiere tables from your OPA policy.
> 
> for these kind of policies to work, make sure to set `opa.policy.column-masking-uri` in the opa-plugin configuration
> 
> here is an example bellow

Reviewer 2: ```suggestion
Column masking allows Trino to mask out/obscure the data in some columns for specific users, without
outright denying access. This plugin supports fetching column masks from OPA.

For these policies to be enabled, make sure to set `opa.policy.column-masking-uri` in the opa-plugin configuration.

For instance, a policy configuring column masking may be implemented as below:
```

Author: sorry pushed a update before seing your proposal :/ 
changed have been re-introduced with commit `a714606`

```
       package trino
            import future.keywords.in
            import future.keywords.if
            import future.keywords.contains

            default allow := true

            column_resource := input.action.resource.column
            is_admin {
              input.context.identity.user == "admin"
            }

            columnMask := {"expression": "NULL"} if {
                not is_admin
                column_resource.catalogName == "sample_catalog"
                column_resource.schemaName == "sample_schema"
                column_resource.tableName == "restricted_table"
                column_resource.columnName == "user_phone"
            }

            columnMask := {"expression": "'****' || substring(user_name, -3)"} if {
                not is_admin
                column_resource.catalogName == "sample_catalog"
                column_resource.schemaName == "sample_schema"
                column_resource.tableName == "restricted_table"
                column_resource.columnName == "user_name"
            }
```

> 

Reviewer 2: Similarly to row filtering, I would mention:

- Unlike row filtering, only _a single_ column mask may be returned for a given column
- The same "identity" field may be returned to evaluate column masks under a different identity

Author: +1, proposal added

  }
}
```

> ## RowFilters 

Reviewer: ```suggestion
## Row filtering

```

> Row filtering allows Trino to filter out some rows from the result before

Reviewer: ```suggestion
Row filtering allows Trino to remove some rows from the result before
```

> returning it to the caller, controlling what data different users can see. This

Reviewer: ```suggestion
returning it to the caller, controlling what data different users can see. The
```

Reviewer: Addressed in replacement PR

> plugin supports retrieving filters from OPA.

Reviewer: ```suggestion
plugin supports retrieving filter definitions from OPA by configuring the OPA endpoint for row filter processing with `opa.policy.row-filters-uri`.
```

Reviewer: Addressed in replacement PR

> For these policies to be enabled, make sure to set `opa.policy.row-filters-uri`

Reviewer: Remove this since its now redundant.

Reviewer: Addressed in replacement PR

in the OPA plugin configuration.

> For instance, an OPA policy for row filtering may be implemented as below:

Reviewer: ```suggestion
For example, an OPA policy for row filtering may be defined by the following rego script:
```

Reviewer: Addressed in replacment PR

```rego
  package trino
  import future.keywords.in
  import future.keywords.if
  import future.keywords.contains

  default allow := true

  table_resource := input.action.resource.table
  is_admin {
    input.context.identity.user == "admin"
  }

  rowFilters contains {"expression": "user_type <> 'customer'"} if {
      not is_admin
      table_resource.catalogName == "sample_catalog"
      table_resource.schemaName == "sample_schema"
      table_resource.tableName == "restricted_table"
  }

> ```

Reviewer: Add empty line after code snippet.

The response expected by the plugin is an array of objects, each of them in the
format `{"expression":"clause"}`.

> Each expression essentially behaves like an additional `WHERE` clause.

Reviewer: Detail some more what that means with the rego script in mind .. what does it do?

Reviewer: To be done later

> Also, It is possible for **multiple** row filters to be returned for a single

Reviewer: Reword this .. maybe even work it into the sentence in line 251

Reviewer: Done in replacement PR.

OPA request, they will all be applied.

> About "identity"

Reviewer: Reword this into a sentence and refer to the identity usage in the script and explain what it does there

- each object may contain an identity field. The identity field allows Trino to
  evaluate these row filters under a **different** identity - such that a filter
  can target a column the requesting user cannot see.
 

## Column masking

> Column masking allows Trino to mask out/obscure the data in some columns for

Reviewer: ```suggestion
Column masking allows Trino to obscure data in one or more columns of the result set for
```

> specific users, without outright denying access. This plugin supports fetching

Reviewer: ```suggestion
specific users, without outright denying access. The plugin supports fetching
```

column masks from OPA.

> For these policies to be enabled, make sure to set

Reviewer: Work this into the sentence above like in row filtering

`opa.policy.column-masking-uri` in the opa-plugin configuration.

> For instance, a policy configuring column masking may be implemented as below:

Reviewer: ```suggestion
For example, a policy configuring column masking may be defined by the following rego script:
```

```rego
  package trino
  import future.keywords.in
  import future.keywords.if
  import future.keywords.contains

  default allow := true

  column_resource := input.action.resource.column
  is_admin {
    input.context.identity.user == "admin"
  }

  columnMask := {"expression": "NULL"} if {
      not is_admin
      column_resource.catalogName == "sample_catalog"
      column_resource.schemaName == "sample_schema"
      column_resource.tableName == "restricted_table"
      column_resource.columnName == "user_phone"
  }

  columnMask := {"expression": "'****' || substring(user_name, -3)"} if {
      not is_admin
      column_resource.catalogName == "sample_catalog"
      column_resource.schemaName == "sample_schema"
      column_resource.tableName == "restricted_table"
      column_resource.columnName == "user_name"
  }
```

> Unlike row filtering, only a **single column mask** may be returned for a given

Reviewer: Same as above .. explain what the script does some more and refactor the section here and following into normal sentences.
